crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,252 @@
+                            {
+                                Ok(Some(new_checkpoint)) => new_checkpoint,
+                                Ok(None) => {
+                                    self.metrics.checkpoint_exec_errors.inc();
+                                    panic!(
+                                        "Checkpoint {:?} does not exist in checkpoint store",
+                                        seq_num,
+                                    );
+                                }
+                                Err(err) => {
+                                    error!("Failed to fetch checkpoint {:?}: {:?}", seq_num, err);
+                                    self.metrics.checkpoint_exec_errors.inc();
+                                    continue;
+                                }
+                            };
+
+                            // Since StateSync is guaranteed to produce checkpoint sync messages
+                            // in monotonically increasing order, then a mismatch of epoch would
+                            // mean that we failed to execute reconfig at the epoch boundary.
+                            assert!(next_checkpoint.epoch() == self.authority_state.epoch());
+
+                            let state = self.authority_state.clone();
+                            let store = self.checkpoint_store.clone();
+                            let next_checkpoint_clone = next_checkpoint.clone();
+                            let metrics = self.metrics.clone();
+                            highest_scheduled_checkpoint = Some(seq_num);
+
+                            finished.push_back(spawn_monitored_task!(async move {
+                                while let Err(err) = execute_checkpoint(
+                                    next_checkpoint_clone.clone(),
+                                    state.clone(),
+                                    store.clone(),
+                                ).await {
+                                    error!(
+                                        "Error while executing checkpoint, will retry in 1s: {:?}",
+                                        err
+                                    );
+                                    tokio::time::sleep(Duration::from_secs(1)).await;
+                                    metrics.checkpoint_exec_errors.inc();
+                                }
+
+                                (seq_num, next_checkpoint_clone)
+                            }));
+
+                            // Last checkpoint of epoch
+                            if let Some(next_committee) = next_checkpoint.next_epoch_committee() {
+                                // Unwrap ok because reconfig failure is unrecoverable
+                                self.reconfig(next_committee).await.unwrap();
+                            }
+
+                            if finished.len() >= task_limit {
+                                break;
+                            }
```

Comment:
While this does prevent us from scheduling more tasks that our limit, spawning tasks this way, inside of the select branch based on notifications from state-sync, means that we may be slower than needed to schedule checkpoints for execution. If our task queue is full but we have 100+ checkpoints to schedule, we probably would want to schedule the next checkpoint as soon as possible once a slot opens up in the queue. The way this is written today would mean that slot would remain open until state-sync happened to notify you of a new checkpoint.

One option would be moving the enqueueing of new tasks to after the select macro but still inside of the `loop`. This way ever time a slot opens we'll have the ability to try and enqueue another one.

Reply:
This ended up being a bit trickier than that. I ended up solving for this edge case by having the select branch that processes finished tasks do an immediate reschedule. So essentially we have the `recv` branch which attempts to schedule as many tasks as possible to saturate the task buffer, then you have the `finished` branch which handles finished branches by ratcheting the watermark and immediately scheduling a single new task to take its place.